### ISSUE TITLE
Affordable Hosting Options for Multiple Terabytes of Data....

### ISSUE THREAD

**Issue by afiaka87** [lucidrains/DALLE-pytorch#113]
Edit: I was perhaps being a bit hyperbolic with my initial posts. I'm still working on this, but it will definitely be something I need an indeterminate amount of time on. 

I'm currently in the process of taking all of the datasets we get our hands on in https://github.com/lucidrains/DALLE-pytorch/discussions/109

Ultimately my plan is to downscale all of them to 256px, and maybe 128px as well and then upload them for us to all download as needed.

I've looked into a few options for this, but essentially with the standard solution (S3 buckets) I can't afford it. Now cloudflare is pretty awesome and will place their CDN in front of your S3 bucket which can save you a ton of traffic. _Especially_ with images as those are small enough to be cached for awhile across their whole CDN, and I can just crank the TTL to a week or something, as the pictures wont change too much after the first upload. 

Honestly, though my preferred option would be academic torrents. On that note, I own a seedbox through leaseweb with 10 gigabit down/up speeds and (essentially) unlimited bandwidth. I could _easily_ just be the first and only stable seeder for awhile. Hopefully others help out and seed as well, but I don't think it will be necessary until we have a ton of people using it.

I could also host them myself there, but if that thing goes down (it does sometimes) you'd all be unable to download. This way if we have two seeders and I go down for a minute, it won't be too bad. 

Does anyone have experience with creating torrent files and starting to seed them on academictorrents? I'm ashamed to admit I've never actually made my own torrent. If so, I would appreciate any info on how I'd go about setting up torrents for all this stuff and hopefully also hosting the torrent files somewhere stable like academictorrents (or I guess just in the repo would work too?)

**Comment by robvanvolt** [comment]
Awesome! I could assist in the first seeds when my new Rig arrives (in the next few weeks)! Also, I am going to ask my university big data center for advice, they usually support scientific big data research projects - maybe we can get a(n) (almost) free hosting for people who "apply" with a simple form for our dataset!

https://www.wikihow.com/Create-a-Torrent to me seems the best instruction for creating a torrent.

Unfortunately I haven't crea

**Comment by afiaka87** [comment]
Okay cool I'm working on getting the scraped images from OpenAI (creative commons I think?) up as a test torrent. Once I've got that I'll probably just use torrent for everything.

**Comment by awilson9** [comment]
Unfortunately I also do not have experience creating torrent files but am happy to seed if you figure it out. My computer isn't running all the time and will have to watch my network usage (I accidentally ate my entire month's data cap in 3 days running wandb in online mode the first few training attempts) but the more seeders the better. And I think just linking the torrents in the readme should suffice

**Comment by robvanvolt** [comment]
About how many TB will the compressed 256x256px (and 158x158px) dataset have? Maybe I will be looking for a big harddrive between 4 - 16 TB I guess?

Also, did we find the best compression parameters? Is 75% jpeg the compression to go? https://pypi.org/project/yfcc100m/ uses 85% e.g. ? Apart from that, I found Joachim Folz' compression script quite inspiring, as he also dealt with aspect ratios and color subsampling ( https://gitlab.com/jfolz/yfcc100m/-/blob/master/yfcc100m/convert.py ) - maybe you find some useful compression hints here. :)

**Comment by afiaka87** [comment]
> About how many TB will the compressed 256x256px (and 158x158px) dataset have? Maybe I will be looking for a big harddrive between 4 - 16 TB I guess?
> 
> Also, did we find the best compression parameters? Is 75% jpeg the compression to go? https://pypi.org/project/yfcc100m/ uses 85% e.g. ? Apart from that, I found Joachim Folz' compression script quite inspiring, as he also dealt with aspect ratios and color subsampling ( https://gitlab.com/jfolz/yfcc100m/-/blob/master/yfcc100m/convert.py ) - maybe you find some useful compression hints here. :)

@robvanvolt
Honestly, I just put terrabytes in the title for good measure. So far everything (after resizing and mild compression) will fit under a terrabyte. The SSD i'm working off of is only a terrabyte and dealing with this stuff on an HDD isn't very fun (the pre-processing stuff I mean - I'm sure it'll run through a data-loader just fine). At any rate, I kinda have to keep things under a terabyte to keep my sanity. 

tl;dr My SSD is currently filled with 800 GiB of data. That's for both the zip of each dataset, the unzipped version of each dataset, and _also_ my resizing data. 

Collective captions is the biggest so far. I got it from 400 GiB down to 50 GiB using gzip compression level 9 after jpeg compression of 75% and resizing to ~256 (they have non-square aspect ratios). Hopefully gives you something of an idea. Sorry I can't give you a full estimate currently. It's alot of stuff ha.

I'm still working on this by the way, slowly but surely. My bottleneck is the amount of time it takes to downsize/convert these things and then my lousy 30Mbps upload speeds at the home internet.

**Comment by robvanvolt** [comment]
Awesome, so I guess my soon-to-arrive 1TB SSD will be enough for now! :D Also, you probably will have all the 'annoying' preprocessing/downloading/uploading stuff done until my rig arrives, thanks for that a lot!!! ^^'

**Comment by jp-krow** [comment]
👀 @afiaka87 
I had no idea you were THIS committed..

I just uploaded the Mega.NZ link to COCO 2018 dataset btw, since google drive isn't such a great option for the long term:
https://mega.nz/file/HDoAQZwQ#j_koJRIw_BkfVhNfW-T2zVhEEiISo_A1fGxozRbW--s

**Comment by afiaka87** [comment]
> 👀 @afiaka87 
> I had no idea you were THIS committed..
> 
> I just uploaded the Mega.NZ link to COCO 2018 dataset btw, since google drive isn't such a great option for the long term:
> https://mega.nz/file/HDoAQZwQ#j_koJRIw_BkfVhNfW-T2zVhEEiISo_A1fGxozRbW--s

Thanks for this! I had to take a mental health day yesterday but I'm gonna start looking at this again in a bit.
